mops/src/opsa/opsa.cu

```diff
@@ -1 +1,393 @@
+                 */
+                for (int32_t i = 0; i < TA; i++) {
+                    if (global_A + i * nThreadRow + threadRow < nfeatures_A) {
+                        regA[i] = A[edge * nfeatures_A + global_A + i * nThreadRow + threadRow];
+                    }
+                }
+
+                /*
+                 *  load B from GMEM into local registers
+                 */
+                for (int32_t i = 0; i < TB; i++) {
+                    if (global_B + i * WARP_SIZE + threadCol < nfeatures_B) {
+                        regB[i] = B[edge * nfeatures_B + global_B + i * WARP_SIZE + threadCol];
+                    }
+                }
+
+                /*
+                 * perform outer product in registers
+                 */
+                for (int32_t i = 0; i < TA; i++) {
+                    for (int32_t j = 0; j < TB; j++) {
+                        regOP[i * TB + j] += regA[i] * regB[j];
+                    }
+                }
+            }
+
+            /*
+             * writeout the content of regOP to the output for this block of
+             * [node, nfeatures_A, nfeatures_B]
+             */
+            for (int32_t j = 0; j < TB; j++) {
+                if (global_B + j * WARP_SIZE + threadCol < nfeatures_B) {
+                    for (int32_t i = 0; i < TA; i++) {
+                        if (global_A + i * nThreadRow + threadRow < nfeatures_A) {
+                            output
+                                [node_index * nfeatures_B * nfeatures_A +
+                                 (global_A + i * nThreadRow + threadRow) * nfeatures_B + global_B +
+                                 j * WARP_SIZE + threadCol] = regOP[i * TB + j];
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+template <typename scalar_t>
+void mops::cuda::outer_product_scatter_add(
+    Tensor<scalar_t, 3> output,
+    Tensor<scalar_t, 2> A,
+    Tensor<scalar_t, 2> B,
+    Tensor<int32_t, 1> indices_output
+) {
+
+    check_sizes(A, "A", 0, B, "B", 0, "opsa");
+    check_sizes(A, "A", 1, output, "output", 1, "opsa");
+    check_sizes(B, "B", 1, output, "output", 2, "opsa");
+    check_sizes(A, "A", 0, indices_output, "indices_output", 0, "opsa");
+
+    int32_t nedges = A.shape[0];
+    int32_t nnodes = output.shape[0];
+    int32_t nfeatures_A = output.shape[1];
+    int32_t nfeatures_B = output.shape[2];
+
+    int32_t *first_occurences = calculate_first_occurences_cuda(indices_output.data, nedges, nnodes);
+
+    dim3 gridDim(nnodes, 1, 1);
+
+    dim3 blockDim(NWARPS_PER_BLOCK * WARP_SIZE, 1, 1);
+
+    outer_product_scatter_add_kernel<scalar_t, 2, 2><<<gridDim, blockDim, 0>>>(
+        A.data,
+        B.data,
+        nnodes,
+        nedges,
+        nfeatures_A,
+        nfeatures_B,
+        first_occurences,
+        indices_output.data,
+        output.data
+    );
+
+    CUDA_CHECK_ERROR(cudaGetLastError());
```

Comment:
not sure I understand this call: does `cudaGetLastError()` return a non-zero status code itself if the previous operation failed?

Reply:
Ok!